### PR TITLE
#189: Add JSON Support to where clause

### PR DESCRIPTION
**Pull request by YahyaHaq**
- Added JSON Support to where clause
- We can now query nested JSON fields
- Added unit and integration tests
- Performed Benchmarking, Here are the becnhmark results [JsonQueryBenchmark.txt](https://github.com/user-attachments/files/16645736/JsonQueryBenchmark.txt)

### CODE REVIEW

File: core/executor.go [hunk 1]
```diff
@@ -215,12 +219,69 @@ func getExprValueAndType(expr sqlparser.Expr, row DSQLQueryResultRow) (interface
 			return nil, "", fmt.Errorf("unknown column: %s", expr.Name.String())
 		}
 	case *sqlparser.SQLVal:
+		// we currently treat JSON query expression as a string value so we will need to differentiate between JSON and SQL strings
+		if expr.Type == sqlparser.StrVal && strings.HasPrefix(string(expr.Val), "_") && getEncoding(row.Value.TypeEncoding) == OBJ_ENCODING_JSON && getType(row.Value.TypeEncoding) == OBJ_TYPE_JSON {
```
Comment (JyotinderSingh) [inline core/executor.go:223]: You can probably add a comment explaining why we're checking for an underscore prefix here.

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]
Please add the benchmarks and unit tests. You can then mark the PR ready for review.

**Comment by JyotinderSingh** [comment]
Thanks for adding the tests @YahyaHaq. Could you please rebase your branch on the latest master to resolve the conflicts?

**Comment by YahyaHaq** [comment]
Yes I'm doing that right now. Also uploading the benchmarks

**Review by JyotinderSingh — APPROVED** [review]
Thanks for adding this feature @YahyaHaq, this will greatly improve the capabilities of our SQL Execution engine and make it a lot more feature-complete.
Thanks!

**Comment by AshwinKul28** [comment]
@YahyaHaq Please check the warning mentioned in the lint check and follow the instructions given to fix it. Thanks

**Comment by JyotinderSingh** [comment]
> @YahyaHaq Please check the warning mentioned in the lint check and follow the instructions given to fix it. Thanks

Will merge this for now and fix the warnings myself as an exception. I am travelling tomorrow so I won't be able to check the PRs.

**Comment by YahyaHaq** [comment]
@JyotinderSingh @arpitbbhayani Thank you so much for merging. This was a great first issue for me and I learned quite a lot about the code base. For the next steps there are some features i think would really improve DiceDB:
1. Adding support for boolean values both in JSON Querying and SET, GET operations
2. Currently whenever we want to evaluate a qwatch query we go through all the keys present and check if they match the key regex provided. I am assuming there are going to be cases where the user provides an exact key that they wants to watch. In that scenario it would be overkill to loop over all the keys stored in our keypool. How about we check that if the user has not provided regex in the key string then we do an exact match from our keypool. This might make our qwatch operation faster for this usecase
3. Fixing the QWATCH bug that i identified that integers are being returned as strings. 

Let me know what you think about these points and i will start to create PRs for these as now i have a strong idea on how to go by this

**Comment by JyotinderSingh** [comment]
> @JyotinderSingh @arpitbbhayani Thank you so much for merging. This was a great first issue for me and I learned quite a lot about the code base. For the next steps there are some features i think would really improve DiceDB:
> 
> 1. Adding support for boolean values both in JSON Querying and SET, GET operations
> 
> 2. Currently whenever we want to evaluate a qwatch query we go through all the keys present and check if they match the key regex provided. I am assuming there are going to be cases where the user provides an exact key that they wants to watch. In that scenario it would be overkill to loop over all the keys stored in our keypool. How about we check that if the user has not provided regex in the key string then we do an exact match from our keypool. This might make our qwatch operation faster for this usecase
> 
> 3. Fixing the QWATCH bug that i identified that integers are being returned as strings. 
> 
> 
> 
> Let me know what you think about these points and i will start to create PRs for these as now i have a strong idea on how to go by this

Amazing to hear this @YahyaHaq! I agree, the issues you mentioned would be good next steps. Could you create the issues for (1) and (3)? (2) is a good suggestion, let's discuss further on it on our discord.

Thanks!

**Pull request merged**